code/weapon/weapons.cpp

```diff
@@ -7054,45 +7054,66 @@ int weapon_area_calc_damage(object *objp, vec3d *pos, float inner_rad, float out
 }
 
 /**
- * Apply the blast effects of an explosion to a ship
+ * Apply the blast effects of an explosion to an object
  *
  * @param force_apply_pos	World pos of where force is applied to object
  * @param ship_objp			Object pointer of ship receiving the blast
  * @param blast_pos			World pos of blast center
  * @param blast				Force of blast
  * @param make_shockwave	Boolean, whether to create a shockwave or not
  */
-void weapon_area_apply_blast(vec3d * /*force_apply_pos*/, object *ship_objp, vec3d *blast_pos, float blast, int make_shockwave)
+void weapon_area_apply_blast(vec3d * /*force_apply_pos*/, object *objp, vec3d *blast_pos, float blast, bool make_shockwave)
 {
 	vec3d		force, vec_blast_to_ship, vec_ship_to_impact;
 	polymodel		*pm;
 
+	Assertion(objp->type == OBJ_SHIP || objp->type == OBJ_ASTEROID || objp->type == OBJ_DEBRIS, "weapon_area_apply_blast can only be called on ships asteroids or debris");
+	if (!(objp->type == OBJ_SHIP || objp->type == OBJ_ASTEROID || objp->type == OBJ_DEBRIS))
+		return;
+
 	// don't waste time here if there is no blast force
 	if ( blast == 0.0f )
 		return;
 
 	// apply blast force based on distance from center of explosion
-	vm_vec_sub(&vec_blast_to_ship, &ship_objp->pos, blast_pos);
+	vm_vec_sub(&vec_blast_to_ship, &objp->pos, blast_pos);
 	vm_vec_normalize_safe(&vec_blast_to_ship);
 	vm_vec_copy_scale(&force, &vec_blast_to_ship, blast );
 
-	vm_vec_sub(&vec_ship_to_impact, blast_pos, &ship_objp->pos);
+	vm_vec_sub(&vec_ship_to_impact, blast_pos, &objp->pos);
+
+	int model_num = -1;
+	switch (objp->type) {
+		case OBJ_SHIP: 
+			model_num = Ship_info[Ships[objp->instance].ship_info_index].model_num;
+			break;
+		case OBJ_ASTEROID:
+			model_num = Ship_info[Ships[objp->instance].ship_info_index].model_num;
+			break;
+		case OBJ_DEBRIS:
+			model_num = Ship_info[Ships[objp->instance].ship_info_index].model_num;
+			break;
```

Comment:
Er, the models for asteroids and debris are using the wrong arrays/vectors here.

You could use `object_get_model` if you want.

Reply:
Whoops. Thanks, I looked for a function like that but I couldn't find it